metals/src/main/scala/scala/meta/internal/metals/codeactions/ExtractRenameMember.scala

```diff
@@ -0,0 +1,419 @@
+
+        val definitions = membersDefinitions(tree)
+        val sldNames = sealedNames(tree)
+        val defnAtCursor =
+          definitions.find(_.name.pos.toLSP.overlapsWith(range))
+
+        def canRenameDefn(defn: Member): Boolean = {
+          val differentNames = defn.name.value != fileName
+          val newFileUri = newPathFromClass(uri, defn)
+
+          differentNames && !newFileUri.exists && defnAtCursor.exists(
+            _.equals(defn)
+          )
+        }
+
+        def canExtractDefn(defn: Member): Boolean = {
+          val differentNames = defn.name.value != fileName
+          val notExtendsSealedOrSealed = notSealed(defn, sldNames)
+          val companion = definitions.find(c => {
+            !c.equals(defn) && c.name.value.equals(defn.name.value)
+          })
+          val companionNotSealed =
+            companion.exists(notSealed(_, sldNames)) || companion.isEmpty
+
+          val newFileUri = newPathFromClass(uri, defn)
+
+          differentNames && notExtendsSealedOrSealed && companionNotSealed && !newFileUri.exists
+        }
+
+        definitions match {
+          case Nil => Nil
+          case head :: Nil if canRenameDefn(head) =>
+            Seq(renameFileAsMemberAction(uri, head))
+          case _ =>
+            val codeActionOpt = for {
+              defn <- defnAtCursor
+              if canExtractDefn(defn)
+              memberType <- getMemberType(defn)
+              title = ExtractRenameMember.title(memberType, defn.name.value)
+            } yield extractClassAction(uri, defn, title)
+
+            codeActionOpt.toList
+        }
+
+      case _ => Nil
+    }
+
+  }
+
+  private def membersDefinitions(tree: Tree): List[Member] = {
+    val nodes: ListBuffer[Member] = ListBuffer()
+
+    val traverser = new SimpleTraverser {
+      override def apply(tree: Tree): Unit = tree match {
+        case p: Pkg =>
+          super.apply(p)
+        case c: Defn.Class => nodes += c
+        case t: Defn.Trait => nodes += t
+        case o: Defn.Object => nodes += o
+        case e: Defn.Enum => nodes += e
+        case s: Source =>
+          super.apply(s)
+        case _ =>
+      }
+    }
+    traverser(tree)
+
+    nodes.toList
+  }
+
+  private def isSealed(t: Tree): Boolean = t match {
+    case node: Defn.Trait => node.mods.exists(_.isInstanceOf[Mod.Sealed])
+    case node: Defn.Class => node.mods.exists(_.isInstanceOf[Mod.Sealed])
+    case _ => false
+  }
+
+  private def sealedNames(tree: Tree): List[String] = {
+    def completeName(node: Member): String = {
+      def completePreName(node: Tree): List[String] = {
+        node.parent match {
+          case Some(t) =>
+            t match {
+              case o: Defn.Object => o.name.value :: completePreName(o)
+              case po: Pkg.Object => po.name.value :: completePreName(po)
+              case _: Source => Nil
+              case _ => completePreName(t)
```

Comment:
That would amount to:
```suggestion
              case _ => Nil
```
no?